src/replication.c

```diff
@@ -190,6 +198,77 @@ void rebaseReplicationBuffer(long long base_repl_offset) {
     }
 }
 
+/* Replication: Primary side - connections association.
+ * On rdb-connection sync, connection association is used to keep replication data in
+ * the backlog until the replica requests PSYNC. Association happens in two forms,
+ * if there's an existing buffer block at the fork time, the replica is attached
+ * to the tail, if there is no tail, the replica will be attached when a new
+ * buffer block is created (see the Retrospect function below).
+ * Replica rdb client id is used as a unique key for the association.
+ * On COB overrun, association is deleted and the RDB connection
+ * is dropped.
+ */
+void addReplicaToPsyncWait(client *replica) {
+    listNode *ln = NULL;
+    replBufBlock *tail = NULL;
+    if (server.repl_backlog == NULL) {
+        createReplicationBacklog();
+    } else {
+        ln = listLast(server.repl_buffer_blocks);
+        tail = ln ? listNodeValue(ln) : NULL;
+        if (tail) {
+            tail->refcount++;
+        }
+    }
+    serverLog(LL_DEBUG, "Add replica %s to waiting psync rax, with cid %llu, %s ", replicationGetReplicaName(replica),
+              (unsigned long long)replica->id, tail ? "with repl-backlog tail" : "repl-backlog is empty");
+    replica->ref_repl_buf_node = tail ? ln : NULL;
+    /* Prevent rdb client from being freed before psync is established. */
+    replica->flags |= CLIENT_PROTECTED_RDB_CONN;
+    uint64_t id = htonu64(replica->id);
+    raxInsert(server.replicas_waiting_psync, (unsigned char *)&id, sizeof(id), replica, NULL);
+}
+
+/* Attach waiting psync replicas with new replication backlog head. */
+void addReplicaToPsyncWaitRetrospect(void) {
+    listNode *ln = listFirst(server.repl_buffer_blocks);
+    replBufBlock *head = ln ? listNodeValue(ln) : NULL;
+    raxIterator iter;
+
+    if (head == NULL) return;
+    /* Update waiting psync replicas to wait on new buffer block */
+    raxStart(&iter, server.replicas_waiting_psync);
+    raxSeek(&iter, "^", NULL, 0);
+    while (raxNext(&iter)) {
+        client *replica = iter.data;
+        if (replica->ref_repl_buf_node) continue;
+        replica->ref_repl_buf_node = ln;
+        head->refcount++;
+        serverLog(LL_DEBUG, "Retrospect attach replica %llu to repl buf block", (long long unsigned int)replica->id);
+    }
+    raxStop(&iter);
+}
+
+void removeReplicaFromPsyncWait(client *replica) {
```

Comment:
What do we call the "normal" replica? `psync_replica` or `main_replica`?
```suggestion
void removeRdbReplicaFromPsyncWait(client *psync_replica) {
```

Reply:
We call the connection rdb-connection and main-connection so I guess main-replica is more accurate, although a bit confusing cause it sounds like there are two replicas. 
What about `replica_rdb_client` and `replica_main_client`?